cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+    def _executor(self) -> AsyncioExecutor:
+        # We must re-use a single Executor due to multi-threading issues in gRPC
+        # clients: https://github.com/grpc/grpc/issues/25364.
+        return AsyncioExecutor.instance()
+
+    # TODO(#5996) Add stop signal to the request iterator.
+    async def _request_iterator(self) -> AsyncIterator[quantum.QuantumRunStreamRequest]:
+        """The request iterator for Quantum Engine client RPC quantum_run_stream().
+
+        Every call to this method generates a new iterator.
+        """
+        while True:
+            yield await self._request_queue.get()
+
+    async def _manage_stream(self):
+        """The stream coroutine, an asyncio coroutine to manage QuantumRunStream.
+
+        This coroutine reads responses from the stream and forwards them to the ResponseDemux, where
+        the corresponding execution coroutine `_manage_request()` is notified.
+
+        When the stream breaks, the stream is reopened, and all execution coroutines are notified.
+
+        There is at most a single instance of this coroutine running.
+        """
+        while True:
+            try:
+                # The default gRPC client timeout is used.
+                response_iterable = await self._grpc_client.quantum_run_stream(
+                    self._request_iterator()
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
+                self._response_demux.publish_exception(e)  # Raise to all request tasks
+                if isinstance(e, asyncio.CancelledError):
+                    break
+
+    async def _manage_execution(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> Union[quantum.QuantumResult, quantum.QuantumJob]:
+        """The execution coroutine, an asyncio coroutine to manage the lifecycle of a job execution.
+
+        This coroutine sends QuantumRunStream requests to the request iterator and receives
+        responses from the ResponseDemux.
+
+        It initially sends a CreateQuantumProgramAndJobRequest, and retries if there is a retryable
+        error by sending another request. The exact request type depends on the error.
+
+        There is one execution coroutine per running job submission.
+        """
+
+        # Construct requests ahead of time to be reused for retries.
+        create_program_and_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        create_job_request = quantum.QuantumRunStreamRequest(
```

Comment:
Does this get used as a retry request type somewhere? I didn't find it ctrl+f'ing around.

Reply:
It's used in `_decide_retry_request_or_raise()` when there is a `StreamError`.